src/osd/OSD.cc

```diff
@@ -7951,31 +7951,19 @@ void OSD::trim_maps(epoch_t oldest, bool skip_maps)
   if (min <= superblock.oldest_map)
     return;
 
-  int num = 0;
   ObjectStore::Transaction t;
-  for (epoch_t e = superblock.oldest_map; e < min; ++e) {
-    dout(20) << " removing old osdmap epoch " << e << dendl;
-    t.remove(coll_t::meta(), get_osdmap_pobject_name(e));
-    t.remove(coll_t::meta(), get_inc_osdmap_pobject_name(e));
-    superblock.oldest_map = e + 1;
-    num++;
-    if (num >= cct->_conf->osd_target_transaction_size) {
-      service.publish_superblock(superblock);
-      write_superblock(cct, superblock, t);
-      int tr = store->queue_transaction(service.meta_ch, std::move(t), nullptr);
-      ceph_assert(tr == 0);
-      num = 0;
-      if (!skip_maps) {
-	// skip_maps leaves us with a range of old maps if we fail to remove all
-	// of them before moving superblock.oldest_map forward to the first map
-	// in the incoming MOSDMap msg. so we should continue removing them in
-	// this case, even we could do huge series of delete transactions all at
-	// once.
-	break;
-      }
-    }
-  }
-  if (num > 0) {
+  // Trim from the superblock's oldest_map up to `min`.
+  // Break if we have exceeded the txn target size.
+  // If skip_maps is true, we will trim up `min` unconditionally.
+  while ((superblock.oldest_map < min) &&
+         (skip_maps ||
+          t.get_num_ops() <= cct->_conf->osd_target_transaction_size)) {
+    dout(20) << " removing old osdmap epoch " << superblock.oldest_map << dendl;
+    t.remove(coll_t::meta(), get_osdmap_pobject_name(superblock.oldest_map));
+    t.remove(coll_t::meta(), get_inc_osdmap_pobject_name(superblock.oldest_map));
```

Comment:
We don't actually want to create transactions of unbounded size.  If I'm understanding the commit message correctly, you are objecting to reusing t after

```
      int tr = store->queue_transaction(service.meta_ch, std::move(t), nullptr);
```

Strictly speaking, I think you are correct that it is UB -- the constructor definition in Transaction.h simply relies on the std::list etc move constructors to leave the source object in a default constructed state.  That doesn't appear to be guaranteed.  How about we create a Transaction::claim_and_reset() method which actually has that guarantee?

Reply:
> If I'm understanding the commit message correctly, you are objecting to reusing t after ..

Yes

> How about we create a Transaction::claim_and_reset() method which actually has that guarantee?

I ended up switching to a transaction vector since `queue_transactions` accepts a txn vector as a parameter.